### PR TITLE
Implement list_tabledata_iter()

### PR DESCRIPTION
**Pull request by krlmlr**
The "iterator" pattern for tabledata, extracted from `list_tabledata_callback()` (which is now using this function).

### CODE REVIEW

File: R/tabledata.r [hunk 1]
```diff
@@ -61,50 +62,80 @@ list_tabledata_callback <- function(project, dataset, table, callback,
...
-  data <- extract_data(req$rows, schema)
-  callback(data)
-  if (!is.null(data)) {
-    rows_fetched <- rows_fetched + nrow(data)
-  }
-  is_complete <- function(rows_fetched) rows_fetched >= as.integer(req$totalRows)
+  cur_page <- 0L
+
+  while(cur_page < max_pages && !iter$is_complete()) {
+    if (iter$is_complete())
```
Comment (craigcitro) [inline R/tabledata.r]: didn't we just check this on the previous line?

File: R/tabledata.r [hunk 2]
```diff
@@ -61,50 +62,80 @@ list_tabledata_callback <- function(project, dataset, table, callback,
...
+    table)
+
+  req <<- NULL
+  rows_fetched <- 0L
+
+  next_ <- function(n) {
+    query <- list(maxResults = n)
+    query$pageToken <- req$pageToken
+
+    # Record only page token and total number of rows to reduce memory consmption
```
Comment (craigcitro) [inline R/tabledata.r]: nit: consumption

File: R/tabledata.r [hunk 3]
```diff
@@ -61,50 +62,80 @@ list_tabledata_callback <- function(project, dataset, table, callback,
...
+    req <- bq_get(url, query = query)
```
Comment (craigcitro) [inline R/tabledata.r]: maybe `response` instead of `request`?

File: R/tabledata.r [hunk 4]
```diff
@@ -61,50 +62,80 @@ list_tabledata_callback <- function(project, dataset, table, callback,
...
+    req <<- req[c("pageToken", "totalRows")]
```
Comment (craigcitro) [inline R/tabledata.r]: don't we want to pull out `req$rows` first? (maybe i'm missing something?)

File: R/tabledata.r [hunk 5]
```diff
@@ -61,50 +62,80 @@ list_tabledata_callback <- function(project, dataset, table, callback,
...
+#' if all rows have been fetched).
+#' @rdname list_tabledata
+#' @export
+list_tabledata_iter <- function(project, dataset, table, callback,
+                                table_info = NULL) {
+
+  table_info <- table_info %||% get_table(project, dataset, table)
+  schema <- table_info$schema
+
+  url <- sprintf("projects/%s/datasets/%s/tables/%s/data", project, dataset,
```
Comment (craigcitro) [inline R/tabledata.r]: overkill for this PR, but maybe worth pulling out a function that takes a table and gives back that url (well, without the `/data` part)?

**Comment by krlmlr** [comment]
Thanks for the review. Removed the unneeded check, now using two variables "response" and "last_response".

**Comment by craigcitro** [comment]
LGTM other than the travis failure

**Comment by krlmlr** [comment]
Should be good now, forgot to document.

**Pull request merged**